merkle-tree/merkle_tree.go

```diff
@@ -0,0 +1,225 @@
+package merkle
+
+import (
+	"bytes"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"hash"
+	"math"
+
+	"github.com/0xPolygon/polygon-edge/crypto"
+	"github.com/0xPolygon/polygon-edge/types"
+)
+
+// A Merkle tree example:
+//
+//              ROOT
+//            /       \
+//          h3         h7
+//         /  \      /    \
+//       h1   h2    h5    h6
+//      / \   / \   / \   / \
+//     L0 L1 L2 L3 L4 L5 L6 L7
+//
+// Each leaf node (L0 - L7) contains a data value, and each intermediate node
+// (h1 - h7) contains the hash of its two child nodes. The root node (ROOT) contains
+// the hash of the entire tree. To prove the inclusion of a leaf node in the tree,
+// a Merkle proof would consist of the leaf node, along with the hash values of its
+// sibling nodes on the path from the leaf to the root.
+//
+// So the proof for leaf node L0 will look like this:
+// H(Hash(L1), Hash(H2), Hash(H7))
+
+var (
+	errLeafNotFound = errors.New("leaf not found")
+)
+
+// MerkleNode represents a single node in merkle tree
+type MerkleNode struct {
+	left   *MerkleNode
+	right  *MerkleNode
+	parent *MerkleNode
+	data   []byte
+	hash   []byte
+}
+
+// newMerkleNode creates a new merkle node
+func newMerkleNode(left, right *MerkleNode, data []byte, hasher hash.Hash) *MerkleNode {
+	node := MerkleNode{left: left, right: right, data: data}
+
+	var dataToHash []byte
+	if left == nil && right == nil {
+		// it's a leaf node
+		dataToHash = data
+	} else {
+		// it's an inner node
+		dataToHash = append(left.hash, right.hash...)
+	}
+
+	hasher.Reset()
+	hasher.Write(dataToHash)
+	node.hash = hasher.Sum(nil)
+
+	return &node
+}
+
+// MerkleTree is the structure for the Merkle tree.
+type MerkleTree struct {
+	// hasher is a pointer to the hashing struct (e.g., Keccak256)
+	hasher hash.Hash
+	// rootNode is the root node of the tree
+	rootNode *MerkleNode
+	// leaf nodes is the list of leaf nodes of the tree (the lowest level of the tree)
+	leafNodes []*MerkleNode
+}
+
+// NewMerkleTree creates a new Merkle tree from the provided data and using the default hashing (Keccak256).
+func NewMerkleTree(data [][]byte) (*MerkleTree, error) {
+	return NewMerkleTreeWithHashing(data, crypto.NewKeccakState())
+}
```

Comment:
Did we say we would get rid off this function in favor of `NewMerkleTreeWithHashing` so that we don't depend on crypto package anymore? Is it the time to do it now or we are going to consider it when transferring logic into a separate library?